File: services/dispute/src/main/java/com/example/iter/dispute/service/ReportService.java
package com.example.iter.dispute.service;

import com.example.iter.auth.api.UserLockPort;
import com.example.iter.auth.api.UserLockView;
import com.example.iter.auth.api.UserQueryPort;
import com.example.iter.auth.api.UserSummary;
import com.example.iter.common.security.UserStatus;
import com.example.iter.common.dto.response.PageResponse;
import com.example.iter.common.exception.CustomException;
import com.example.iter.common.exception.ErrorCode;
import com.example.iter.dispute.domain.entity.Report;
import com.example.iter.dispute.domain.entity.ReportStatus;
import com.example.iter.dispute.domain.repository.ReportRepository;
import com.example.iter.dispute.domain.repository.spec.ReportSpecifications;
import com.example.iter.dispute.dto.request.ReportCreateRequest;
import com.example.iter.dispute.dto.request.ReportSearchRequest;
import com.example.iter.dispute.dto.response.ReportDetailResponse;
import com.example.iter.dispute.dto.response.ReportSummaryResponse;
import com.example.iter.dispute.util.ReportMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.EnumSet;
import java.util.List;
import java.util.Set;

@Slf4j
@Service
@RequiredArgsConstructor
public class ReportService {

    private static final Set<ReportStatus> ACTIVE_REPORT_STATUSES = EnumSet.of(
            ReportStatus.RECEIVED,
            ReportStatus.UNDER_REVIEW
    );

    private final ReportRepository reportRepository;
    private final UserQueryPort userQueryPort;
    private final UserLockPort userLockPort;
    private final ReportTargetValidator reportTargetValidator;
    private final ReportMapper reportMapper;

    // 신고 대상과 처리 중인 중복 신고를 검증한 뒤 신고를 접수합니다.
    @Transactional
    public ReportDetailResponse createReport(Long reporterId, ReportCreateRequest request) {
        lockAndValidateReporter(reporterId);

        reportTargetValidator.validate(
                request.targetType(),
                request.targetId(),
                reporterId
        );

        validateDuplicateActiveReport(reporterId, request);

        Report report = Report.builder()
                .reporterId(reporterId)
                .targetType(request.targetType())
                .targetId(request.targetId())
                .reason(request.reason().trim())
                .description(request.description().trim())
                .status(ReportStatus.RECEIVED)
                .build();

        Report savedReport = reportRepository.save(report);
        log.info("신고 접수 처리: reportId={}, reporterId={}, targetType={}, targetId={}, status={}",
                savedReport.getId(), reporterId, savedReport.getTargetType(),
                savedReport.getTargetId(), savedReport.getStatus());

        // 락은 "신고해도 되는 회원인가"를 묻고, 닉네임은 응답 표시용이라 조회가 따로다.
        // 검증에 실패하면 여기까지 오지 않으므로 실패 경로에서는 쿼리가 늘지 않는다.
        UserSummary reporter = userQueryPort.findSummary(reporterId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        return reportMapper.toDetail(savedReport, reporter);
    }

    // 로그인 사용자가 작성한 신고를 검색 조건과 페이지 정보로 조회합니다.
    @Transactional(readOnly = true)
    public PageResponse<ReportSummaryResponse> getMyReports(Long reporterId, ReportSearchRequest request) {
        UserSummary reporter = userQueryPort.findSummary(reporterId).orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        Page<ReportSummaryResponse> reports = reportRepository.findAll(
                ReportSpecifications.myReports(reporterId, request.targetType(), request.status()),
                reportPageable(request.page(), request.size())
        ).map(report -> reportMapper.toSummary(report, reporter));

        return PageResponse.from(reports);
    }

    // 로그인 사용자가 작성한 특정 신고의 상세 정보를 조회합니다.
    @Transactional(readOnly = true)
    public ReportDetailResponse getMyReport(Long reporterId, Long reportId) {
        UserSummary reporter = userQueryPort.findSummary(reporterId).orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));
        Report report = reportRepository.findByIdAndReporterId(reportId, reporterId).orElseThrow(() -> new CustomException(ErrorCode.REPORT_NOT_FOUND));

        return reportMapper.toDetail(report, reporter);
    }

    // 신고 생성 동시 요청을 직렬화하고 신고자가 현재 이용 가능한 상태인지 검증합니다.
    private void lockAndValidateReporter(Long reporterId) {
        UserLockView reporter = userLockPort.lockAll(List.of(reporterId)).get(reporterId);
        if (reporter == null) {
            throw new CustomException(ErrorCode.USER_NOT_FOUND);
        }

        // 차단 사유와 에러 코드는 신고 정책이 정한다. 포트는 상태만 돌려준다.
        if (reporter.status() == UserStatus.SUSPENDED) {
            throw new CustomException(ErrorCode.USER_SUSPENDED);
        }

        if (reporter.status() == UserStatus.DELETED) {
            throw new CustomException(ErrorCode.USER_DELETED);
        }
    }

    // 같은 사용자의 같은 대상 신고가 이미 처리 중이면 신고 생성을 제한합니다.
    private void validateDuplicateActiveReport(Long reporterId, ReportCreateRequest request) {
        boolean duplicateExists = reportRepository.existsActiveReport(
                reporterId,
                request.targetType(),
                request.targetId(),
                ACTIVE_REPORT_STATUSES
        );

        if (duplicateExists) {
            throw new CustomException(ErrorCode.DUPLICATE_ACTIVE_REPORT);
        }
    }

    // 신고 목록을 최신 접수 순으로 조회할 페이지 정보를 생성합니다.
    private Pageable reportPageable(int page, int size) {
        return PageRequest.of(
                page,
                size,
                Sort.by(
                        Sort.Order.desc("createdAt"),
                        Sort.Order.desc("id")
                )
        );
    }
}

File: services/dispute/src/main/kotlin/com/example/iter/dispute/service/ReportService.kt
package com.example.iter.dispute.service

import com.example.iter.auth.api.UserLockPort
import com.example.iter.auth.api.UserQueryPort
import com.example.iter.common.dto.response.PageResponse
import com.example.iter.common.exception.CustomException
import com.example.iter.common.exception.ErrorCode
import com.example.iter.common.security.UserStatus
import com.example.iter.dispute.domain.entity.Report
import com.example.iter.dispute.domain.entity.ReportStatus
import com.example.iter.dispute.domain.repository.ReportRepository
import com.example.iter.dispute.domain.repository.spec.ReportSpecifications
import com.example.iter.dispute.dto.request.ReportCreateRequest
import com.example.iter.dispute.dto.request.ReportSearchRequest
import com.example.iter.dispute.dto.response.ReportDetailResponse
import com.example.iter.dispute.dto.response.ReportSummaryResponse
import com.example.iter.dispute.util.ReportMapper
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ReportService(
    private val reportRepository: ReportRepository,
    private val userQueryPort: UserQueryPort,
    private val userLockPort: UserLockPort,
    private val reportTargetValidator: ReportTargetValidator,
    private val reportMapper: ReportMapper
) {

    // 신고자 행을 먼저 잠근 뒤 대상 정책과 처리 중 중복 여부를 확인하여 신고를 접수합니다.
    @Transactional
    fun createReport(reporterId: Long, request: ReportCreateRequest): ReportDetailResponse {
        lockAndValidateReporter(reporterId)

        val targetType = requireNotNull(request.targetType)
        val targetId = requireNotNull(request.targetId)

        reportTargetValidator.validate(targetType, targetId, reporterId)
        validateDuplicateActiveReport(reporterId, targetType, targetId)

        val savedReport = reportRepository.save(
            Report(
                reporterId = reporterId,
                targetType = targetType,
                targetId = targetId,
                reason = requireNotNull(request.reason).trim(),
                description = requireNotNull(request.description).trim(),
                status = ReportStatus.RECEIVED,
            )
        )

        log.info(
            "신고 접수 처리: reportId={}, reporterId={}, targetType={}, targetId={}, status={}",
            savedReport.id,
            reporterId,
            savedReport.targetType,
            savedReport.targetId,
            savedReport.status
        )

        val reporter = userQueryPort.findSummary(reporterId).orElseThrow { CustomException(ErrorCode.USER_NOT_FOUND) }

        return reportMapper.toDetail(savedReport, reporter)
    }

    // 본인 신고만 조건 검색하며 기존 외부 계약에 맞춰 오프셋 페이지 응답을 유지합니다.
    @Transactional(readOnly = true)
    fun getMyReports(reporterId: Long, request: ReportSearchRequest): PageResponse<ReportSummaryResponse> {
        val reporter = userQueryPort.findSummary(reporterId).orElseThrow { CustomException(ErrorCode.USER_NOT_FOUND) }

        val reports = reportRepository.findAll(
            ReportSpecifications.myReports(reporterId, request.targetType, request.status),
            reportPageable(request.page, request.size)
        ).map { report -> reportMapper.toSummary(report, reporter) }

        return PageResponse.from(reports)
    }

    // reportId와 reporterId를 함께 조회 조건으로 사용해 타인의 신고 존재 여부도 노출하지 않습니다.
    @Transactional(readOnly = true)
    fun getMyReport(reporterId: Long, reportId: Long): ReportDetailResponse {
        val reporter = userQueryPort.findSummary(reporterId).orElseThrow { CustomException(ErrorCode.USER_NOT_FOUND) }

        val report = reportRepository.findByIdAndReporterId(reportId, reporterId).orElseThrow { CustomException(ErrorCode.REPORT_NOT_FOUND) }

        return reportMapper.toDetail(report, reporter)
    }

    // 같은 신고자가 보낸 동시 요청을 직렬화하여 애플리케이션 중복 검사 사이의 경쟁을 줄입니다.
    private fun lockAndValidateReporter(reporterId: Long) {
        val reporter = userLockPort.lockAll(listOf(reporterId))[reporterId]
            ?: throw CustomException(ErrorCode.USER_NOT_FOUND)

        when (reporter.status) {
            UserStatus.SUSPENDED -> throw CustomException(ErrorCode.USER_SUSPENDED)
            UserStatus.DELETED -> throw CustomException(ErrorCode.USER_DELETED)
            else -> Unit
        }
    }

    // 접수·검토 중인 같은 대상 신고만 중복으로 판단하고 종결 후 재신고는 허용합니다.
    private fun validateDuplicateActiveReport(reporterId: Long, targetType: com.example.iter.dispute.domain.entity.ReportTargetType, targetId: Long) {
        if (reportRepository.existsActiveReport(reporterId, targetType, targetId, ACTIVE_REPORT_STATUSES)) {
            throw CustomException(ErrorCode.DUPLICATE_ACTIVE_REPORT)
        }
    }

    private fun reportPageable(page: Int, size: Int): Pageable = PageRequest.of(
        page,
        size,
        Sort.by(Sort.Order.desc("createdAt"), Sort.Order.desc("id"))
    )

    private companion object {
        private val ACTIVE_REPORT_STATUSES = setOf(ReportStatus.RECEIVED, ReportStatus.UNDER_REVIEW)
        private val log = LoggerFactory.getLogger(ReportService::class.java)
    }
}
